### ISSUE TITLE
feat(`terraform_providers_lock`): Run only when non-cross-platform lockfile found

### ISSUE THREAD

**Issue by MaxymVlasov** [antonbabenko/pre-commit-terraform#526]
### What problem are you facing?

Running `terraform_providers_lock` for each provider for each Architecture tooks about 2 seconds.
We have 4 arches (`linux amd64`, `linux arm64`, `mac amd64`, `mac arm64`) and up to 14 providers.
That took about 2-3 minutes to check the lockfile, which it too much.

I found that during `terraform init`, lockfile not touched in case all providers already here. Each provider block looks like:

 ```hcl
provider "registry.terraform.io/okta/okta" {
  version     = "3.46.0"
  constraints = ">= 3.0.0, ~> 3.38"
  hashes = [
    "h1:8AQNT9HErs8LdGbkU+X7U7ggJtbZuJ15hLIfG+tNxzs=",
    "h1:Ab6Zjnf0jzF8GAk11QLRvOTp37DaavzNLQEpVePZVKg=",
    "h1:TtoJ+D96IIxr4660CGZeTd3VQQCkGZHHgSylhsy/B2A=",
    "h1:efq7Hv5KDfuIIffVSi541msYTcv4HmID5xzVO8ZJzdc=",
    "zh:033e9c282bfb1a47be6701b2d56da14c9cb214079f445adf0a3cf2be3bf546e4",
    "zh:0a49da96635c42cc118e507c171e2fb228d8789a9eeaf3de4b01cd47a624751a",
    "zh:1103e29b144b6f5ec607a32988bc490b661909b86ed2c37d09a08e8cff32a99c",
    "zh:824cdf825f7bf6b367213969f167d39bc0b27fcd5bff56031ee720f89df79714",
    "zh:a8f1f19698c6e0d4b53655d5c55c821398805549047c493e264cdfdf58d9b053",
    "zh:aefd9fc1d88ad2e4990dc52ed470f5d61f115b071897bc91fb5e54311d3599ea",
    "zh:c501aad6dbeeb262757a235ae386d59f758dabc609d545d2c766b4b5ce951859",
    "zh:c67a663e796f425fb1daa6281e726c025aa7199d50c7e93603d4a853ca56282d",
    "zh:d0e2af10e773063b587a2d3acf2317d1eacb68f091b85b124c05dab88a6966bf",
    "zh:de06df4e14817049acbb800c09de702708557cae1a0727f4248b18ff927df664",
    "zh:e551113f1e26c641ab55a6fe4250a23a71afd19719b5ff0b32cda425af5461ca",
    "zh:e5b0c9ab19876a072b21335fe059624684bb7c54317ab598404cc3af7f5faf37",
    "zh:e991a7c70472f987e6d6f4a12ddea91de0875c9b839f1cf61cfe13e27d9a35ea",
    "zh:fc992ddcee35a364904b675bb17642fe80df3c3dcf8674dbc6527e2af58b419b",
  ]
}
```

In the case there is no provider in lockfile yet, the next block will be added by `terraform init`:

```hcl
provider "registry.terraform.io/okta/okta" {
  version     = "3.46.0"
  constraints = ">= 3.0.0, ~> 3.38"
  hashes = [
    "h1:efq7Hv5KDfuIIffVSi541msYTcv4HmID5xzVO8ZJzdc=",
  ]
}
```

After run `terraform providers lock`:
```hcl
provider "registry.terraform.io/okta/okta" {
  version     = "3.46.0"
  constraints = ">= 3.0.0, ~> 3.38"
  hashes = [
    "h1:efq7Hv5KDfuIIffVSi541msYTcv4HmID5xzVO8ZJzdc=",
    "zh:033e9c282bfb1a47be6701b2d56da14c9cb214079f445adf0a3cf2be3bf546e4",
    "zh:0a49da96635c42cc118e507c171e2fb228d8789a9eeaf3de4b01cd47a624751a",
    "zh:1103e29b144b6f5ec607a32988bc490b661909b86ed2c37d09a08e8cff32a99c",
    "zh:824cdf825f7bf6b367213969f167d39bc0b27fcd5bff56031ee720f89df79714",
    "zh:a8f1f19698c6e0d4b53655d5c55c821398805549047c493e264cdfdf58d9b053",
    "zh:aefd9fc1d88ad2e4990dc52ed470f5d61f115b071897bc91fb5e54311d3599ea",
    "zh:c501aad6dbeeb262757a235ae386d59f758dabc609d545d2c766b4b5ce951859",
    "zh:c67a663e796f425fb1daa6281e726c025aa7199d50c7e93603d4a853ca56282d",
    "zh:d0e2af10e773063b587a2d3acf2317d1eacb68f091b85b124c05dab88a6966bf",
    "zh:de06df4e14817049acbb800c09de702708557cae1a0727f4248b18ff927df664",
    "zh:e551113f1e26c641ab55a6fe4250a23a71afd19719b5ff0b32cda425af5461ca",
    "zh:e5b0c9ab19876a072b21335fe059624684bb7c54317ab598404cc3af7f5faf37",
    "zh:e991a7c70472f987e6d6f4a12ddea91de0875c9b839f1cf61cfe13e27d9a35ea",
    "zh:fc992ddcee35a364904b675bb17642fe80df3c3dcf8674dbc6527e2af58b419b",
  ]
}
```

---

At the same time, if version in `terraform {}` block changed - lockfile regenerates too, during `terraform init`

IE, was:

```hcl
provider "registry.terraform.io/okta/okta" {
  version     = "3.45.0"
  constraints = ">= 3.0.0, 3.45.0"
  hashes = [
    "h1:ulUT2eziHbyZEQoLFCR1ZEDqUgO/1kHUa6PYHruIx3Y=",
  ]
}
```
after changing `constraints`

```hcl
provider "registry.terraform.io/okta/okta" {
  version     = "3.46.0"
  constraints = ">= 3.0.0, ~> 3.38"
  hashes = [
    "h1:efq7Hv5KDfuIIffVSi541msYTcv4HmID5xzVO8ZJzdc=",
  ]
}
```

---

If provider removed - it automatically removed from lockfile during `terraform init`
 
### How could pre-commit-terraform help solve your problem?

I propose, adding a flag like `--check-only-for-cross-platform` or `--do-not-force-update`. And when the flag is specified - trigger the `terraform providers lock` command only when no `zh:` will be found at least for one provider.

Or compare `constraints` in lockfile and in tf-code.

That should save time and make you able to use this hook in complex root modules.

**Comment by yermulnik** [comment]
To be fair there should be no such cases where this file has to be generated via `pre-commit` every time `pre-commit-terraform` is run. This file should be committed into VCS once there are changes to providers or their version are introduced.
So I'd say you're workarounding something very local/custom to your pipeline.

You also should not relay on hash identities as they may change w/o prior notification:
> We may occasionally introduce new hashing schemes if we learn of limitations in the existing schemes or if a new scheme offers some considerable additional benefit.

You may refer to https://developer.hashicorp.com/terraform/language/files/dependency-lock#new-provider-package-checksums to get more info on TF deps lock file hash schemes and what are they.

**Comment by MaxymVlasov** [comment]
>To be fair there should be no such cases where this file has to be generated via pre-commit every time pre-commit-terraform is run.

That's how it works now

**Comment by yermulnik** [comment]
> That's how hook works now ¯\_(ツ)_/¯

Simply don't lock provider if `.terraform.lock.hcl` exists?
```shell
if [ -r .terraform.lock.hcl -a -s .terraform.lock.hcl ]; then
  common::colorify 'yellow' 'Terraform provider lock file exists. Skipping its creation.'
  return
else
  terraform providers lock "${args[@]}"
fi

**Comment by MaxymVlasov** [comment]
.terraform.lock.hcl always exist, if you at least once run `terraform init`

`--skip-providers-lock` - it just "don't use terraform_providers_lock hook" option :)

**Comment by antonbabenko** [comment]
This issue has been resolved in version 1.80.0 :tada:
